src/pointer/ptr.rs

```diff
@@ -1050,7 +1064,7 @@ mod _casts {
             //   `UnsafeCell<T>` has the same in-memory representation as its
             //   inner type `T`. A consequence of this guarantee is that it is
             //   possible to convert between `T` and `UnsafeCell<T>`.
-            let ptr = unsafe { ptr.assume_validity::<I::Validity>() };
+            let ptr = unsafe { ptr.assume_validity::<V>() };
```

Comment:
Maybe introduce a `Validity::WithValidity<V>` GAT that permits going from `V<Inner = UnsafeCell<T>>` to `W<Inner = UnsafeCell<T>>`?